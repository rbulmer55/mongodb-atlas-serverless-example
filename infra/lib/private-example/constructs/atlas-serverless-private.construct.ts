import { config, Environment } from '@infra/config';
import { StackProps } from 'aws-cdk-lib';
import { SubnetType, Vpc } from 'aws-cdk-lib/aws-ec2';
import {
	AtlasServerlessBasic,
	CfnServerlessPrivateEndpoint,
	ScopeDefinitionType,
	ServerlessInstanceProviderSettingsProviderName,
} from 'awscdk-resources-mongodbatlas';
import { Construct } from 'constructs';

interface AtlasStackProps {
	readonly profile: string;
	readonly region: string;
	readonly instanceName: string;
	readonly projectName: string;
	readonly continuousBackupEnabled: boolean;
	readonly terminationProtectionEnabled: boolean;
	readonly username: string;
}

interface CustomPrivateAtlasServerlessProps extends StackProps {
	atlas: {
		orgId: string;
		vpc: Vpc;
	};
}

export class PrivateAtlasServerless extends Construct {
	public readonly vpceId: string;
	constructor(
		scope: Construct,
		id: string,
		props: CustomPrivateAtlasServerlessProps
	) {
		super(scope, id);

		const atlasProps = this.getContextProps();

		const atlas = new AtlasServerlessBasic(this, 'PrivateAtlasSlsInstance', {
			serverlessProps: {
				name: atlasProps.instanceName,
				profile: atlasProps.profile,
				continuousBackupEnabled: atlasProps.continuousBackupEnabled,
				providerSettings: {
					providerName:
						ServerlessInstanceProviderSettingsProviderName.SERVERLESS,
					regionName: atlasProps.region,
				},
				terminationProtectionEnabled: atlasProps.terminationProtectionEnabled,
			},
			projectProps: {
				name: atlasProps.projectName,
				orgId: props.atlas.orgId,
			},
			dbUserProps: {
				username: atlasProps.username,
				scopes: [
					{ name: atlasProps.instanceName, type: ScopeDefinitionType.CLUSTER },
				],
			},
			ipAccessListProps: {
				accessList: [{ cidrBlock: props.atlas.vpc.vpcCidrBlock }],
			},
			profile: atlasProps.profile,
		});

		/**
		 * MAY/24 - RB
		 * Notes:
		 *
		 * Issues with the Security Group - Must use the default security group with PrivateLink and Atlas Serverless (See Below).
		 *
		 * CfnPrivateEndpointService - This Atlas CloudFormation resource does not support a serverless instance
		 *
		 * CfnServerlessPrivateEndpoint - This resource does not allow you to change the security group for the generated VPCE.
		 * 								If using just to create the EndpointService, you are unable to get the EndpointServiceId OR VPCE Id to add a security group in CDK.
		 *
		 * CfnPrivateEndpointAws - This resource would be used to link up the Atlas EndpointService and the VPCE, however as we can't get the EndpointService Id yet we can't use this.
		 */
		const cfnPrivateEndpoint = new CfnServerlessPrivateEndpoint(
			this,
			'AtlasSlsPrivateEndpoint',
			{
				profile: atlasProps.profile,
				instanceName: atlasProps.instanceName,
				projectId: atlas.mProject.attrId,
				createAndAssignAwsPrivateEndpoint: true,
				awsPrivateEndpointConfigurationProperties: {
					vpcId: props.atlas.vpc.vpcId,
					subnetIds: props.atlas.vpc.selectSubnets({
						subnetType: SubnetType.PRIVATE_ISOLATED,
					}).subnetIds,
					region: atlasProps.region,
				},
			}
		);
		cfnPrivateEndpoint.addDependency(atlas.mserverless);
	}

	private getContextProps(): AtlasStackProps {
		const profile = config.environment;
		const region = config.region.toUpperCase().replace(/-/gi, '_');
		const instanceName = `${config.environment}-${config.domain}-${config.projectName}`;
		const projectName = `${config.environment}-${config.domain}-${config.projectName}`;
		const username = `${config.environment}-${config.domain}-${config.projectName}-user`;
		const terminationProtectionEnabled =
			config.environment === Environment.Prd ? true : false;
		const continuousBackupEnabled =
			config.environment === Environment.Prd ? true : false;

		return {
			terminationProtectionEnabled,
			continuousBackupEnabled,
			profile,
			region,
			instanceName,
			projectName,
			username,
		};
	}
}
