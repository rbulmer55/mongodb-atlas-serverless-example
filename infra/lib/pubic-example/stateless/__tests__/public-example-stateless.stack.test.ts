import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { PublicExampleStatelessStack } from '@infra/lib/pubic-example/stateless/public-example-stateless.stack';
import { IVpc, SubnetType, Vpc } from 'aws-cdk-lib/aws-ec2';
import { Construct } from 'constructs';
import { PublicExampleStatefulStack } from '../../stateful/public-example-stateful.stack';

const env = {
	domain: 'mydomain',
	environment: 'fdv',
	mongoAtlasOrgId: 'org-abc123',
	projectName: 'atlas-example',
	region: 'eu-west-1',
};
const mockVPC = (construct: Construct): IVpc => {
	return new Vpc(construct, 'mocked-vpc', {
		cidr: '10.0.0.0/16',
		maxAzs: 2,
		subnetConfiguration: [
			{
				cidrMask: 24,
				name: 'pub-subnet',
				subnetType: SubnetType.PUBLIC,
			},
			{
				cidrMask: 28,
				name: 'prv-subnet',
				subnetType: SubnetType.PRIVATE_ISOLATED,
			},
		],
	});
};

export const processTemplate = (template: Template) => {
	return JSON.parse(
		JSON.stringify(template, null, 4).replace(
			/"S3Key":\s".+?\..+?"/gi,
			'"S3Key": "Any<String>"'
		)
	);
};

beforeEach(() => {
	jest.resetModules();
	process.env = {
		REGION: env.region,
		ENVIRONMENT: env.environment,
		DOMAIN_NAME: env.domain,
		PROJECT_NAME: env.projectName,
		MONGO_ORG_ID: env.mongoAtlasOrgId,
	};
});
describe('All atlas-example stacks created and has correct properties', () => {
	test('Public Stateless Stack', () => {
		const app = new App();

		const statefulStack = new PublicExampleStatefulStack(
			app,
			'PublicStatefulStack'
		);

		const stack = new PublicExampleStatelessStack(app, 'PublicStatelessStack', {
			atlasVpc: statefulStack.applicationVpc,
		});

		expect(processTemplate(Template.fromStack(stack))).toMatchSnapshot();
	});
});
