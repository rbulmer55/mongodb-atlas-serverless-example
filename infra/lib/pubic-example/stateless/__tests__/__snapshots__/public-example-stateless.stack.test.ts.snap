// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`All atlas-example stacks created and has correct properties Public Stateless Stack 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "LambdaSecurityGroup0BD9FC99": {
      "Properties": {
        "GroupDescription": "PublicStatelessStack/LambdaSecurityGroup",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "SecurityGroupIngress": [
          {
            "CidrIp": {
              "Fn::ImportValue": "PublicStatefulStack:ExportsOutputFnGetAttAppVpc80F1F7F9CidrBlockDFD56ADE",
            },
            "Description": {
              "Fn::Join": [
                "",
                [
                  "from ",
                  {
                    "Fn::ImportValue": "PublicStatefulStack:ExportsOutputFnGetAttAppVpc80F1F7F9CidrBlockDFD56ADE",
                  },
                  ":ALL TRAFFIC",
                ],
              ],
            },
            "IpProtocol": "-1",
          },
        ],
        "VpcId": {
          "Fn::ImportValue": "PublicStatefulStack:ExportsOutputRefAppVpc80F1F7F93D0AA00C",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": {
      "DependsOn": [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "5bc602ecde93c947efe5899ae355f999986a1acbe610b1c0b9c468d738857555.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": [
          {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "dbconnectionsecretA4B511ED": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "Name": "db/atlas/fdv/atlas-example",
        "SecretString": "{"dbConnectionString":"REPLACE_ME"}",
      },
      "Type": "AWS::SecretsManager::Secret",
      "UpdateReplacePolicy": "Delete",
    },
    "incrementCounterLambda538CADD4": {
      "DependsOn": [
        "incrementCounterLambdaServiceRoleDefaultPolicy7FC45F80",
        "incrementCounterLambdaServiceRoleBCA39A2B",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "933f6964caf466a58f6194bf8f2458da0cfc0c79fece88b695aec08cedfe392c.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "COUNTER_DB_CONNECTION_SECRET": "db/atlas/fdv/atlas-example",
          },
        },
        "Handler": "index.incrementCounter",
        "Role": {
          "Fn::GetAtt": [
            "incrementCounterLambdaServiceRoleBCA39A2B",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 5,
        "VpcConfig": {
          "SecurityGroupIds": [
            {
              "Fn::GetAtt": [
                "LambdaSecurityGroup0BD9FC99",
                "GroupId",
              ],
            },
          ],
          "SubnetIds": [
            {
              "Fn::ImportValue": "PublicStatefulStack:ExportsOutputRefAppVpcfdvmydomainatlasexamplepvtappSubnet1SubnetAB43409231770349",
            },
          ],
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "incrementCounterLambdaLogRetention884FD1FB": {
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "incrementCounterLambda538CADD4",
              },
            ],
          ],
        },
        "RetentionInDays": 90,
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "incrementCounterLambdaServiceRoleBCA39A2B": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaVPCAccessExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "incrementCounterLambdaServiceRoleDefaultPolicy7FC45F80": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "secretsmanager:GetSecretValue",
                "secretsmanager:DescribeSecret",
              ],
              "Effect": "Allow",
              "Resource": {
                "Ref": "dbconnectionsecretA4B511ED",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "incrementCounterLambdaServiceRoleDefaultPolicy7FC45F80",
        "Roles": [
          {
            "Ref": "incrementCounterLambdaServiceRoleBCA39A2B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
